core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1300,24 +1302,34 @@ public NameResolverRegistry getNameResolverRegistry() {
     public void onResult(final ResolutionResult resolutionResult) {
       final class NamesResolved implements Runnable {
 
-        @SuppressWarnings("ReferenceEquality")
+        @SuppressWarnings({"ReferenceEquality", "unchecked"})
         @Override
         public void run() {
           List<EquivalentAddressGroup> servers = resolutionResult.getAddresses();
           Attributes attrs = resolutionResult.getAttributes();
           channelLogger.log(
               ChannelLogLevel.DEBUG, "Resolved address: {0}, config={1}", servers, attrs);
 
+          boolean nrReplied = haveBackends == null;
+
           if (haveBackends == null || !haveBackends) {
             channelLogger.log(ChannelLogLevel.INFO, "Address resolved: {0}", servers);
             haveBackends = true;
           }
 
           nameResolverBackoffPolicy = null;
+          ConfigOrError configOrError = resolutionResult.getServiceConfig();
+          Map<String, ?> serviceConfig = null;
+          if (configOrError != null) {
+            if (configOrError.getConfig() != null) {
+              serviceConfig = (Map<String, ?>) configOrError.getConfig();
+            } else if (nrReplied || defaultServiceConfig == null) {
+              // First DNS lookup has invalid service config, and cannot fall back to default(=null)
+              onError(configOrError.getError());
```

Comment:
This should be down in the else-branch of `if (!lookUpServiceConfig)`. If `lookUpServiceConfig == false`, the service config doesn't matter thus invalid config shouldn't be turned into an error. See my comment below.